### ISSUE TITLE
"make check" fails to build cleanly with extra warnings enabled

### ISSUE THREAD

**Issue by Patater** [ARMmbed/mbedtls#1628]
### Description
- Type: Bug
- Priority: Minor

---------------------------------------------------------------
## Bug

**mbed TLS build:**  
Version: 2.9.0

**Expected behavior** 
`make check CFLAGS="-O2 -Wall -Wextra -Werror"` should not fail to build.

**Actual behavior**  

`make check CFLAGS="-O2 -Wall -Wextra -Werror"` fails to build, due to warnings like the following:
```
  CC    test_suite_cipher.aes.c
helpers.function:224:13: error: unused function 'hexify'
      [-Werror,-Wunused-function]
static void hexify( unsigned char *obuf, const unsigned char *ibuf, int len )
            ^
helpers.function:278:23: error: unused function 'unhexify_alloc'
      [-Werror,-Wunused-function]
static unsigned char *unhexify_alloc( const char *ibuf, size_t *olen )
                      ^
helpers.function:329:12: error: unused function 'rnd_zero_rand'
      [-Werror,-Wunused-function]
static int rnd_zero_rand( void *rng_state, unsigned char *output, size_t len )
           ^
helpers.function:356:12: error: unused function 'rnd_buffer_rand'
      [-Werror,-Wunused-function]
static int rnd_buffer_rand( void *rng_state, unsigned char *output, size_t len )
           ^
helpers.function:402:12: error: unused function 'rnd_pseudo_rand'
      [-Werror,-Wunused-function]
static int rnd_pseudo_rand( void *rng_state, unsigned char *output, size_t len )
           ^
5 errors generated.
make[1]: *** [test_suite_cipher.aes] Error 1
```

**Steps to reproduce**

Run `make check CFLAGS="-O2 -Wall -Wextra -Werror"`.

**Comment by gilles-peskine-arm** [comment]
`make check CFLAGS=-Werror` does the same thing, except that it disables the warnings on unused functions and values (we turn on `-Wall -Wextra` in the default compiler flags). The files `test_suite_*.c` are automatically generated, so it makes sense to generate code for sometimes-used functions and let the compiler eliminate the ones that aren't used in a specific test.

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-2524

**Comment by mpg** [comment]
@Patater I can't reproduce this on my machine. Steps used:
```
make clean
git checkout mbedtls-2.9.0
make check CFLAGS="-O2 -Wall -Wextra -Werror"
```

Tools versions:
```
% gcc --version | head -n1                                                                                                                               
gcc (Ubuntu 5.5.0-12ubuntu1~16.04) 5.5.0 20171010
% make --version | head -n1                                                                                                                              
GNU Make 4.1
```

Can you still reproduce on your machine? If so, any idea what could explain the discrepancy?

Note: we do try to suppress this warning in tests, because as Gilles mentioned, we don't really care about it in generated code.
```
% grep CFLAGS tests/Makefile | uniq
CFLAGS  ?= -O2
WARNING_CFLAGS ?= -Wall -W -Wdeclaration-after-statement -Wno-unused-function -Wno-unused-value
LOCAL_CFLAGS = $(WARNING_CFLAGS) -I../include -D_FILE_OFFSET_BITS=64
LOCAL_CFLAGS += -g3
        $(CC) $(LOCAL_CFLAGS) $(CFLAGS) $<      $(LOCAL_LDFLAGS) $(LDFLAGS) -o $@
```

**Comment by Patater** [comment]
I was using clang.

**Comment by mpg** [comment]
Could reproduce now. It looks like the issue would be fixed by changing each occurrence of `$(LOCAL_CFLAGS) $(CFLAGS)` to `$(CFLAGS) $(LOCAL_CFLAGS)` in `tests/Makefile`.

Or, as Gilles said, in the meantime you can just use `CFLAGS='-Werror -O2'` which does the same as your invocation except it works.

**Comment by sbutcher-arm** [comment]
I've just realised I've created the same bug over again because I encountered it in another way. See #2051.

I've already raised a PR to fix this with PR #2053.

**Comment by yanesca** [comment]
Fixed by https://github.com/ARMmbed/mbedtls/pull/2053, therefore closing.
